src/idiomatic/leveraging-the-type-system/borrow-checker-invariants.md

```diff
@@ -0,0 +1,81 @@
+---
+minutes: 10
+---
+
+# Using the Borrow checker to enforce Invariants
+
+The logic of the borrow checker, while tied to "memory ownership", can be
+abstracted away from this central use case to model other problems and prevent
+API misuse.
+
+```rust,editable
+// Doors can be open or closed, and you need the right key to lock or unlock
+// one. Modelled with a Shared key and Owned door.
+pub struct DoorKey {
+    pub key_shape: u32,
+}
+pub struct LockedDoor {
+    lock_shape: u32,
+}
+pub struct OpenDoor {
+    lock_shape: u32,
+}
+
+fn open_door(key: &DoorKey, door: LockedDoor) -> Result<OpenDoor, LockedDoor> {
+    if door.lock_shape == key.key_shape {
+        Ok(OpenDoor { lock_shape: door.lock_shape })
+    } else {
+        Err(door)
+    }
+}
+
+fn close_door(key: &DoorKey, door: OpenDoor) -> Result<LockedDoor, OpenDoor> {
+    if door.lock_shape == key.key_shape {
+        Ok(LockedDoor { lock_shape: door.lock_shape })
+    } else {
+        Err(door)
+    }
+}
+
+fn main() {
+    let key = DoorKey { key_shape: 7 };
+    let closed_door = LockedDoor { lock_shape: 7 };
+    let opened_door = open_door(&key, closed_door);
+    if let Ok(opened_door) = opened_door {
+        println!("Opened the door with key shape '{}'", key.key_shape);
+    } else {
+        eprintln!(
+            "Door wasn't opened! Your key only opens locks with shape '{}'",
+            key.key_shape
+        );
+    }
+}
+```
+
+<details>
+
+<!-- TODO: link to typestate when that gets merged. -->
+
+- The borrow checker has been used to prevent use-after-free and multiple
+  mutable references up until this point, and we've used types to shape and
+  restrict use of APIs already using
+  [the Typestate pattern](../leveraging-the-type-system/typestate-pattern.md).
```

Comment:
Please break up complex sentences into multiple simpler sentences that are easier to scan for the instructor while speaking. Something like:

- We've seen the borrow checker prevent memory safety bugs (use-after-free, data races).

- We've also used the typestate pattern to restrict use of APIs on objects in inappropriate states.